Start Pasers Combinators lesson (#44)
haskellfoundation/HaskellSchool: en/lessons/parsers/introduction.md

---
version: 1.0.0
title: Introduction to Parser Combinators
---

{% include toc.html %}

> Parsers are ubiquitous in Haskell litterature because, thanks to Parser Combinators it illustrates the power of composition.

Reviewer: ```suggestion
Parsers are ubiquitous in educational Haskell literature because parser combinators can be used to illustrate the power of composition.
```

Correct me if I'm wrong about the implication here being that parser combinators are a useful tool in teaching Haskell. 

I have replaced the pronouns to make the sentence a little clearer (in general however, "they" is used to replace a plural subject).

> Parsers, while being mostly associated to compilers and interpreters, can be found everywhere.

Reviewer: ```suggestion
Parsers, while being mostly associated with compilers and interpreters, can be found everywhere.
```

> Their goal is to take a input data (usually a text or a stream), and produce a structured output (usually a data structure, such as an [AST](https://en.wikipedia.org/wiki/Abstract_syntax_tree)), see [this article](https://lexi-lambda.github.io/blog/2019/11/05/parse-don-t-validate/) for day to day application.

Reviewer: ```suggestion
Their goal is to take input data (usually text or a stream), and produce structured output (usually a data structure, such as an [AST](https://en.wikipedia.org/wiki/Abstract_syntax_tree)). See [this article](https://lexi-lambda.github.io/blog/2019/11/05/parse-don-t-validate/) for day to day application.
```
I have removed "a" a couple times here: I believe your language, especially when referring to types, was grammatical, but in my opinion this sounds more natural. 

I also replaced a comma with a period to avoid the grammatical issue of comma splicing. 

This is likely a later consideration in the review process, but it seems a good idea to get permission from the author to link the article and, if this occurs, to ask her how she wants the article clarified.

Author: Sure, I'll contact the author, thanks.

Parser Combinators are a way to build parsers by composition.

> Let's say we want to extract all the tuples of a text, we can start by defining our tuple parser:

Reviewer: ```suggestion
Let's say we want to extract all the tuples from some text. We can start by defining our tuple parser:
```

These suggestions are similar to the previous suggestion.

```haskell
import Text.ParserCombinators.ReadP

tuple :: ReadP (String, String)
tuple =
  between (char '(') (char ')') $ do
    left <- munch1 (/= ',')
    char ','
    skipSpaces
    right <- munch1 (/= ')')
    return (left, right)
```

> As you can see, relying on `Monad`, allows us to focus on the structure on the input, instead of the structure on the Parser.

Reviewer: ```suggestion
As you can see, relying on `Monad` allows us to focus on the structure of the input, instead of the structure of the parser.
```
The word "parser" usually doesn't get capitalised in English.

> then, we can define our main parser, composition:

Reviewer: I think the comma here gets in the way of clarity.

```haskell
tuples :: ReadP [(String, String)]
tuples = many tuple
```

> that's the power of Parser Combinator, since `ReadP` implements `Alternative`, any parser can be reused and extended (here via [`many`](https://hackage.haskell.org/package/base/docs/Control-Applicative.html#v:many)).

Reviewer: ```suggestion
This is the power of parser combinators: since `ReadP` implements `Alternative`, any parser can be reused and extended (here via [`many`](https://hackage.haskell.org/package/base/docs/Control-Applicative.html#v:many)).
```
"This" may sound more natural here: "this" refers to things that have been mentioned very recently. 

The change in punctuation was suggested to remove a comma splice.

We can have a toy example:

```haskell
example :: [([(String, String)], String)]
example = readP_to_S tuples "(abc,def)(gh,   ij)(k,l)"
```

which gives us:

```haskell
[
    ([], "(abc,def)(gh,   ij)(k,l)")
,   ([("abc", "def") ], "(gh,   ij)(k,l)")
,   ([("abc", "def"), ("gh", "ij")], "(k,l)")
,   ([( "abc", "def"), ("gh", "ij"), ("k", "l")] , "")
]
```

> As you can see, running `ReadP` gives all the possible parsed result, which can lead to performances issues.

Reviewer: ```suggestion
As you can see, running `ReadP` gives all possible parsed results, which can lead to performances issues.
```
